WalletWasabi.Tests/UnitTests/Blockchain/TransactionBuilding/BranchAndBoundTests.cs

```diff
@@ -13,84 +14,175 @@ public class BranchAndBoundTests
 	/// </summary>
 	[Fact]
-	public void TargetIsBiggerThanBalance()
+	public void ExactMatch_TargetIsBiggerThanBalance()
 	{
-		long target = 5_000;
 		List<long> inputValues = new();
+		long target = 5_000;
 
 		for (int i = 0; i < target - 1; i++)
 		{
 			inputValues.Add(1);
 		}
 
-		BranchAndBound selector = new(inputValues);
-		bool wasSuccessful = selector.TryGetExactMatch(target, out List<long>? selectedCoins);
+		BranchAndBound algorithm = new();
+		ExactMatchStrategy strategy = new(target, inputValues.ToArray());
+		bool wasSuccessful = algorithm.TryGetMatch(strategy, out List<long>? selectedCoins);
 
 		Assert.False(wasSuccessful);
 		Assert.Null(selectedCoins);
 	}
 
 	[Fact]
-	public void ReturnNullIfNoExactMatchFoundTest()
+	public void ExactMatch_ReturnNullIfNoExactMatchFoundTest()
 	{
-		List<long> inputValues = new() { 120_000, 100_000, 100_000, 50_000, 40_000 };
-		BranchAndBound selector = new(inputValues);
+		long[] inputValues = new long[] { 120_000, 100_000, 100_000, 50_000, 40_000 };
 		long target = 300000;
 
-		bool wasSuccessful = selector.TryGetExactMatch(target, out List<long>? selectedCoins);
+		BranchAndBound algorithm = new();
+		ExactMatchStrategy strategy = new(target, inputValues);
+		bool wasSuccessful = algorithm.TryGetMatch(strategy, out List<long>? selectedCoins);
+
+		Assert.False(wasSuccessful);
+		Assert.Null(selectedCoins);
+	}
+
+	/// <summary>Tests that a best found selection is found when an exact solution does not exist.</summary>
+	[Fact]
+	public void CheapestSelection_NoInputCosts()
+	{
+		long[] inputValues = new long[] { 11, 7, 5, 3, 2 }; // Sum is 28.
+		long[] inputCosts = new long[] { 0, 0, 0, 0, 0 }; // No input costs. Idealized.
+		long target = 27; // Target that we cannot get as a sum of input values.
+
+		BranchAndBound algorithm = new();
+		CheapestSelectionStrategy strategy = new(target, inputValues, inputCosts);
+		bool wasSuccessful = algorithm.TryGetMatch(strategy, out List<long>? selectedCoins);
 
 		Assert.False(wasSuccessful);
 		Assert.Null(selectedCoins);
+		Assert.Equal(new long[] { 11, 7, 5, 3, 2 }, strategy.GetBestSelectionFound());
 	}
 
-	private List<long> GenerateListOfRandomValues(int count = 1000)
+	[Fact]
+	public void CheapestSelection_ExactMatchIsAlsoCheapest()
+	{
+		long[] inputValues = new long[] { 35, 17, 10, 5, 3, 2 };
+
+		// Make the second input very expensive to spend so that it is not selected (not likely in reality).
+		long[] inputCosts = new long[] { 1, 5, 1, 1, 1, 1 };
+
+		long target = 27;
+
+		BranchAndBound algorithm = new();
+		CheapestSelectionStrategy strategy = new(target, inputValues, inputCosts);
+		bool wasSuccessful = algorithm.TryGetMatch(strategy, out List<long>? selectedCoins);
+
+		Assert.False(wasSuccessful);
+		Assert.Null(selectedCoins);
+
+		// There are multiple existing solutions.
+		long[][] solutions = new long[][]  {
+			new long[] { 17, 10 },
+			new long[] { 17, 5, 3, 2 },
+			new long[] { 35 }
+		};
```

Comment:
We shouldn't have multiple solutions IMO.

- `new long[] { 17, 10 }` has `6` input cost, so it adds up to `33`. `33 - 27 = 6 burn away`
- `new long[] { 17, 5, 3, 2 }` has `8` input cost, so it adds up to `35`.  `35 - 27 = 8 burn away`
- `new long[] { 35 }` has 1 input cost, so it adds up to `36`.  `36 - 27 = 9 burn away`

Shouldn't the algo only give back `new long[] { 17, 10 }` ?

Reply:
Yes, you are right. I have fixed it. Thanks